File: monitor.js
import 'dotenv/config';
import fs from 'fs';
import { setTimeout } from 'timers/promises';
import fetch from 'node-fetch';

class PurpleAirMonitor {
  constructor() {
    this.apiKey = process.env.PURPLEAIR_API_KEY;
    this.sensorIndexes = [
      '240075', '240049', '239257', '239297', 
      '239307', '239267', '240595', '239259', 
      '240067', '222835', '222887'
    ];
    this.baseUrl = 'https://api.purpleair.com/v1/sensors';
    this.supabaseUrl = process.env.VITE_SUPABASE_URL;
    this.supabaseKey = process.env.VITE_SUPABASE_ANON_KEY;
    
    // Configure logging
    this.logFile = 'sensor_monitor.log';
    this.log('PurpleAir Monitor initialized');
    this.log(`Monitoring ${this.sensorIndexes.length} sensors`);
  }

  log(message, level = 'INFO') {
    const timestamp = new Date().toISOString();
    const logMessage = `${timestamp} - ${level} - ${message}\n`;
    
    // Log to console
    console.log(logMessage.trim());
    
    // Log to file
    fs.appendFileSync(this.logFile, logMessage);
  }

  async getSensorData(sensorIndex) {
    try {
      const url = `${this.baseUrl}/${sensorIndex}`;
      const response = await fetch(url, {
        headers: {
          'X-API-Key': this.apiKey,
          'Content-Type': 'application/json'
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      this.log(`Error fetching data for sensor ${sensorIndex}: ${error.message}`, 'ERROR');
      return null;
    }
  }

  async checkSensorStatus() {
    const currentTime = Date.now() / 1000; // Convert to Unix timestamp
    const offlineSensors = [];

    for (const sensorIndex of this.sensorIndexes) {
      const data = await this.getSensorData(sensorIndex);
      if (!data) {
        this.log(`Failed to get data for sensor ${sensorIndex}`, 'ERROR');
        continue;
      }

      const sensor = data.sensor || {};
      const lastSeen = sensor.last_seen;
      const name = sensor.name || `Sensor ${sensorIndex}`;

      if (lastSeen) {
        const timeDifference = currentTime - lastSeen;
        const hoursOffline = timeDifference / 3600;

        if (hoursOffline > 3) {
          offlineSensors.push({
            name,
            index: sensorIndex,
            hours: Math.round(hoursOffline * 10) / 10,
            lastSeen: new Date(lastSeen * 1000).toISOString()
          });
          this.log(`Sensor ${name} (ID: ${sensorIndex}) hasn't reported in ${Math.round(hoursOffline * 10) / 10} hours`, 'WARNING');
        }
      } else {
        this.log(`No last_seen data for sensor ${sensorIndex}`, 'WARNING');
      }
    }

    if (offlineSensors.length > 0) {
      // Trigger email notification via Edge Function
      try {
        const response = await fetch(`${this.supabaseUrl}/functions/v1/notify`, {
          method: 'POST',
          headers: {
            'Authorization': `Bearer ${this.supabaseKey}`,
            'Content-Type': 'application/json'
          }
        });

        if (!response.ok) {
          throw new Error(`Failed to send notification: ${response.status}`);
        }

        const result = await response.json();
        this.log(`Notification sent successfully: ${JSON.stringify(result)}`, 'INFO');
      } catch (error) {
        this.log(`Failed to send notification: ${error.message}`, 'ERROR');
      }
    }
  }

  async run(checkInterval = 300000) { // 5 minutes default
    this.log('Starting PurpleAir sensor monitoring');
    
    while (true) {
      try {
        await this.checkSensorStatus();
        await setTimeout(checkInterval);
      } catch (error) {
        this.log(`Error in monitoring loop: ${error.message}`, 'ERROR');
        await setTimeout(60000); // Wait a minute before retrying if there's an error
      }
    }
  }
}

// Start the monitor
const monitor = new PurpleAirMonitor();
monitor.run().catch(error => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: monitor.py
import os
import time
import logging
from datetime import datetime, timezone
from urllib.request import Request, urlopen
import json
from urllib.error import URLError
from http.client import HTTPResponse

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('sensor_monitor.log'),
        logging.StreamHandler()
    ]
)

class PurpleAirMonitor:
    def __init__(self):
        self.api_key = os.environ.get('PURPLEAIR_API_KEY')
        self.sensor_indexes = [
            '240075', '240049', '239257', '239297', 
            '239307', '239267', '240595', '239259', 
            '240067', '222835', '222887'
        ]
        self.base_url = 'https://api.purpleair.com/v1/sensors'
        
        logging.info("PurpleAir Monitor initialized")
        logging.info(f"Monitoring {len(self.sensor_indexes)} sensors")

    def get_sensor_data(self, sensor_index):
        try:
            url = f"{self.base_url}/{sensor_index}"
            headers = {
                'X-API-Key': self.api_key,
                'Content-Type': 'application/json'
            }
            
            request = Request(
                url,
                headers=headers,
                method='GET'
            )
            
            with urlopen(request) as response:
                if isinstance(response, HTTPResponse):
                    data = response.read()
                    return json.loads(data)
                return None
                
        except URLError as e:
            logging.error(f"Error fetching data for sensor {sensor_index}: {e}")
            return None
        except Exception as e:
            logging.error(f"Unexpected error for sensor {sensor_index}: {e}")
            return None

    def log_notification(self, subject, message):
        logging.warning("\n=== NOTIFICATION ===")
        logging.warning(f"Subject: {subject}")
        logging.warning(message)
        logging.warning("==================\n")

    def check_sensor_status(self):
        current_time = datetime.now(timezone.utc).timestamp()
        offline_sensors = []

        for sensor_index in self.sensor_indexes:
            data = self.get_sensor_data(sensor_index)
            if not data:
                logging.error(f"Failed to get data for sensor {sensor_index}")
                continue

            sensor = data.get('sensor', {})
            last_seen = sensor.get('last_seen')
            name = sensor.get('name', f'Sensor {sensor_index}')

            if last_seen:
                time_difference = current_time - last_seen
                hours_offline = time_difference / 3600

                if hours_offline > 3:  # Changed threshold to 3 hours
                    offline_sensors.append({
                        'name': name,
                        'index': sensor_index,
                        'hours': round(hours_offline, 1),
                        'last_seen': datetime.fromtimestamp(last_seen, timezone.utc).isoformat()
                    })
                    logging.warning(
                        f"Sensor {name} (ID: {sensor_index}) hasn't reported in {round(hours_offline, 1)} hours"
                    )
            else:
                logging.warning(f"No last_seen data for sensor {sensor_index}")

        if offline_sensors:
            message = "The following sensors have been offline for more than 3 hours:\n\n"
            for sensor in offline_sensors:
                message += f"- {sensor['name']} (ID: {sensor['index']})\n"
                message += f"  Last seen: {sensor['last_seen']}\n"
                message += f"  Hours offline: {sensor['hours']}\n\n"
            
            self.log_notification(
                "PurpleAir Sensors Offline Alert",
                message
            )

    def run(self, check_interval=300):  # 5 minutes default
        logging.info("Starting PurpleAir sensor monitoring")
        while True:
            try:
                self.check_sensor_status()
                time.sleep(check_interval)
            except Exception as e:
                logging.error(f"Error in monitoring loop: {e}")
                time.sleep(60)  # Wait a minute before retrying if there's an error

if __name__ == "__main__":
    monitor = PurpleAirMonitor()
    monitor.run()